File: step_15_graphs/15.5_mst_disjoint_set/P006_accounts_merge/solution.py
"""
Problem: Accounts Merge (LeetCode #721)
Difficulty: MEDIUM | XP: 25

Given a list of accounts where accounts[i] = [name, email1, email2, ...],
merge accounts that share at least one email. Return merged accounts with
emails sorted and the account name prepended. Two accounts belong to the
same person if they share any email.
"""
from typing import List
from collections import defaultdict


# ============================================================
# APPROACH 1: BRUTE FORCE -- Pairwise account comparison
# Time: O(n^2 * m) where n=accounts, m=avg emails per account
# Space: O(n * m)
# For each pair of accounts, check if they share any email.
# If yes, merge them. Repeat until no more merges are possible.
# ============================================================
def brute_force(accounts: List[List[str]]) -> List[List[str]]:
    # Represent each account as a set of emails (index 0 is the name)
    groups = [(acc[0], set(acc[1:])) for acc in accounts]
    merged = True

    while merged:
        merged = False
        new_groups = []
        used = [False] * len(groups)

        for i in range(len(groups)):
            if used[i]:
                continue
            name_i, emails_i = groups[i]
            for j in range(i + 1, len(groups)):
                if used[j]:
                    continue
                name_j, emails_j = groups[j]
                if emails_i & emails_j:  # common email exists
                    emails_i = emails_i | emails_j
                    used[j] = True
                    merged = True
            new_groups.append((name_i, emails_i))
            used[i] = True

        groups = new_groups

    result = []
    for name, emails in groups:
        result.append([name] + sorted(emails))
    return result


# ============================================================
# APPROACH 2: OPTIMAL -- Union-Find (DSU) on account indices
# Time: O(n * m * alpha(n))  |  Space: O(n * m)
# Map each email to the index of the FIRST account that owns it.
# For each subsequent account that also owns that email, union the
# two account indices. Collect all emails per root and sort.
# ============================================================
def optimal(accounts: List[List[str]]) -> List[List[str]]:
    n = len(accounts)
    parent = list(range(n))
    rank = [0] * n

    def find(x: int) -> int:
        if parent[x] != x:
            parent[x] = find(parent[x])  # path compression
        return parent[x]

    def union(x: int, y: int) -> None:
        rx, ry = find(x), find(y)
        if rx == ry:
            return
        if rank[rx] < rank[ry]:
            parent[rx] = ry
        elif rank[rx] > rank[ry]:
            parent[ry] = rx
        else:
            parent[ry] = rx
            rank[rx] += 1

    # Map email -> first account index that owns it
    email_to_account: dict = {}

    for i, acc in enumerate(accounts):
        for email in acc[1:]:
            if email in email_to_account:
                # This email was seen before -> same person -> union
                union(i, email_to_account[email])
            else:
                email_to_account[email] = i

    # Group all emails by root account index
    root_to_emails: dict = defaultdict(set)
    for email, acc_idx in email_to_account.items():
        root = find(acc_idx)
        root_to_emails[root].add(email)

    # Build result: [name, sorted emails...]
    result = []
    for root, emails in root_to_emails.items():
        name = accounts[root][0]
        result.append([name] + sorted(emails))

    return result


# ============================================================
# APPROACH 3: BEST -- Union-Find with email-string DSU (no index)
# Time: O(n * m * alpha(n * m))  |  Space: O(n * m)
# Instead of unioning account indices, union email strings directly.
# Each email's root is mapped to an account name for reconstruction.
# ============================================================
def best(accounts: List[List[str]]) -> List[List[str]]:
    parent: dict = {}

    def find(x: str) -> str:
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]

    def union(x: str, y: str) -> None:
        rx, ry = find(x), find(y)
        if rx != ry:
            parent[rx] = ry

    email_to_name: dict = {}

    for acc in accounts:
        name = acc[0]
        first_email = acc[1]
        # Initialize all emails from this account
        for email in acc[1:]:
            if email not in parent:
                parent[email] = email
            email_to_name[email] = name
            # Union all emails in this account with the first one
            union(first_email, email)

    # Group by root
    root_to_emails: dict = defaultdict(list)
    for email in parent:
        root = find(email)
        root_to_emails[root].append(email)

    result = []
    for root, emails in root_to_emails.items():
        name = email_to_name[root]
        result.append([name] + sorted(emails))

    return result


if __name__ == "__main__":
    print("=== Accounts Merge ===\n")

    accounts1 = [
        ["John","johnsmith@mail.com","john_newyork@mail.com"],
        ["John","johnsmith@mail.com","john00@mail.com"],
        ["Mary","mary@mail.com"],
        ["John","johnnybravo@mail.com"]
    ]

    r1 = brute_force([acc[:] for acc in accounts1])
    r2 = optimal([acc[:] for acc in accounts1])
    r3 = best([acc[:] for acc in accounts1])

    print("Brute:")
    for acc in sorted(r1): print(" ", acc)

    print("\nOptimal:")
    for acc in sorted(r2): print(" ", acc)

    print("\nBest:")
    for acc in sorted(r3): print(" ", acc)

    # Expected:
    # ["John","john00@mail.com","john_newyork@mail.com","johnsmith@mail.com"]
    # ["John","johnnybravo@mail.com"]
    # ["Mary","mary@mail.com"]

File: step_15_graphs/15.5_mst_disjoint_set/P006_accounts_merge/Solution.java
import java.util.*;

/**
 * Problem: Accounts Merge (LeetCode #721)
 * Difficulty: MEDIUM | XP: 25
 *
 * Merge accounts that share at least one email address.
 * Return merged accounts with the account name and sorted email list.
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Pairwise account comparison
    // Time: O(n^2 * m)  |  Space: O(n * m)
    // Repeatedly check every pair of accounts for shared emails.
    // Merge and repeat until no more merges occur (convergence).
    // ============================================================
    public static List<List<String>> bruteForce(List<List<String>> accounts) {
        // Represent each account as (name, Set<email>)
        List<String> names = new ArrayList<>();
        List<Set<String>> emailSets = new ArrayList<>();

        for (List<String> acc : accounts) {
            names.add(acc.get(0));
            Set<String> emails = new HashSet<>(acc.subList(1, acc.size()));
            emailSets.add(emails);
        }

        boolean merged = true;
        while (merged) {
            merged = false;
            List<String> newNames = new ArrayList<>();
            List<Set<String>> newSets = new ArrayList<>();
            boolean[] used = new boolean[emailSets.size()];

            for (int i = 0; i < emailSets.size(); i++) {
                if (used[i]) continue;
                String name = names.get(i);
                Set<String> emails = new HashSet<>(emailSets.get(i));

                for (int j = i + 1; j < emailSets.size(); j++) {
                    if (used[j]) continue;
                    // Check overlap
                    boolean overlap = false;
                    for (String e : emailSets.get(j)) {
                        if (emails.contains(e)) { overlap = true; break; }
                    }
                    if (overlap) {
                        emails.addAll(emailSets.get(j));
                        used[j] = true;
                        merged = true;
                    }
                }
                newNames.add(name);
                newSets.add(emails);
                used[i] = true;
            }
            names = newNames;
            emailSets = newSets;
        }

        List<List<String>> result = new ArrayList<>();
        for (int i = 0; i < names.size(); i++) {
            List<String> sorted = new ArrayList<>(emailSets.get(i));
            Collections.sort(sorted);
            sorted.add(0, names.get(i));
            result.add(sorted);
        }
        return result;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Union-Find on account indices
    // Time: O(n * m * alpha(n))  |  Space: O(n * m)
    // Map each email to the index of the first account that owns it.
    // Union account indices when they share an email. Group emails
    // by root account index and sort.
    // ============================================================
    public static List<List<String>> optimal(List<List<String>> accounts) {
        int n = accounts.size();
        int[] parent = new int[n];
        int[] rank = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;

        // Map email -> first account index that contains it
        Map<String, Integer> emailToAccount = new HashMap<>();

        for (int i = 0; i < n; i++) {
            List<String> acc = accounts.get(i);
            for (int j = 1; j < acc.size(); j++) {
                String email = acc.get(j);
                if (emailToAccount.containsKey(email)) {
                    union(parent, rank, i, emailToAccount.get(email));
                } else {
                    emailToAccount.put(email, i);
                }
            }
        }

        // Group emails by root account index
        Map<Integer, List<String>> rootToEmails = new HashMap<>();
        for (Map.Entry<String, Integer> entry : emailToAccount.entrySet()) {
            int root = find(parent, entry.getValue());
            rootToEmails.computeIfAbsent(root, x -> new ArrayList<>()).add(entry.getKey());
        }

        List<List<String>> result = new ArrayList<>();
        for (Map.Entry<Integer, List<String>> entry : rootToEmails.entrySet()) {
            int root = entry.getKey();
            List<String> emails = entry.getValue();
            Collections.sort(emails);
            emails.add(0, accounts.get(root).get(0)); // prepend name
            result.add(emails);
        }
        return result;
    }

    private static int find(int[] parent, int x) {
        if (parent[x] != x) parent[x] = find(parent, parent[x]); // path compression
        return parent[x];
    }

    private static void union(int[] parent, int[] rank, int x, int y) {
        int rx = find(parent, x), ry = find(parent, y);
        if (rx == ry) return;
        if (rank[rx] < rank[ry]) parent[rx] = ry;
        else if (rank[rx] > rank[ry]) parent[ry] = rx;
        else { parent[ry] = rx; rank[rx]++; }
    }

    // ============================================================
    // APPROACH 3: BEST -- Union-Find on email strings directly
    // Time: O(n * m * alpha(n * m))  |  Space: O(n * m)
    // Union email strings directly using a HashMap-based DSU.
    // Each email's root email maps to the account name.
    // Avoids integer index bookkeeping.
    // ============================================================
    public static List<List<String>> best(List<List<String>> accounts) {
        Map<String, String> parent = new HashMap<>();
        Map<String, String> emailToName = new HashMap<>();

        // Initialize and union within each account
        for (List<String> acc : accounts) {
            String name = acc.get(0);
            String firstEmail = acc.get(1);
            for (int j = 1; j < acc.size(); j++) {
                String email = acc.get(j);
                parent.putIfAbsent(email, email);
                emailToName.put(email, name);
                unionStr(parent, firstEmail, email);
            }
        }

        // Group by root
        Map<String, List<String>> rootToEmails = new HashMap<>();
        for (String email : parent.keySet()) {
            String root = findStr(parent, email);
            rootToEmails.computeIfAbsent(root, x -> new ArrayList<>()).add(email);
        }

        List<List<String>> result = new ArrayList<>();
        for (Map.Entry<String, List<String>> entry : rootToEmails.entrySet()) {
            String root = entry.getKey();
            List<String> emails = entry.getValue();
            Collections.sort(emails);
            emails.add(0, emailToName.get(root));
            result.add(emails);
        }
        return result;
    }

    private static String findStr(Map<String, String> parent, String x) {
        if (!parent.get(x).equals(x))
            parent.put(x, findStr(parent, parent.get(x)));
        return parent.get(x);
    }

    private static void unionStr(Map<String, String> parent, String x, String y) {
        String rx = findStr(parent, x), ry = findStr(parent, y);
        if (!rx.equals(ry)) parent.put(rx, ry);
    }

    public static void main(String[] args) {
        System.out.println("=== Accounts Merge ===\n");

        List<List<String>> accounts = new ArrayList<>(Arrays.asList(
            new ArrayList<>(Arrays.asList("John","johnsmith@mail.com","john_newyork@mail.com")),
            new ArrayList<>(Arrays.asList("John","johnsmith@mail.com","john00@mail.com")),
            new ArrayList<>(Arrays.asList("Mary","mary@mail.com")),
            new ArrayList<>(Arrays.asList("John","johnnybravo@mail.com"))
        ));

        System.out.println("Brute:");
        List<List<String>> r1 = bruteForce(new ArrayList<>(accounts));
        r1.sort(Comparator.comparing(a -> a.toString()));
        r1.forEach(System.out::println);

        System.out.println("\nOptimal:");
        List<List<String>> r2 = optimal(new ArrayList<>(accounts));
        r2.sort(Comparator.comparing(a -> a.toString()));
        r2.forEach(System.out::println);

        System.out.println("\nBest:");
        List<List<String>> r3 = best(new ArrayList<>(accounts));
        r3.sort(Comparator.comparing(a -> a.toString()));
        r3.forEach(System.out::println);
    }
}
